packages/core/src/codewhisperer/nextEditPrediction/PredictionTracker.ts

```diff
@@ -0,0 +1,235 @@
+    }
+
+    /**
+     * Enforces memory limits by removing old snapshots if necessary
+     */
+    private async enforceMemoryLimits(): Promise<void> {
+        while (this.storageSize > this.config.maxStorageSizeKb * 1024) {
+            const oldestFile = this.findOldestFile()
+            if (!oldestFile) {
+                break
+            }
+
+            const fileSnapshots = this.snapshots.get(oldestFile)
+            if (!fileSnapshots || fileSnapshots.length === 0) {
+                this.snapshots.delete(oldestFile)
+                continue
+            }
+
+            const removedSnapshot = fileSnapshots.shift()
+            if (removedSnapshot) {
+                this.storageSize -= removedSnapshot.size
+                this.logger.debug(
+                    `Snapshot deleted (memory limit) for file: ${removedSnapshot.filePath}, remaining snapshots: ${this.getTotalSnapshotCount()}, new size: ${Math.round(this.storageSize / 1024)} KB`
+                )
+            }
+
+            if (fileSnapshots.length === 0) {
+                this.snapshots.delete(oldestFile)
+            }
+        }
+    }
+
+    /**
+     * Finds the file with the oldest snapshot
+     * @returns The file path of the oldest snapshot
+     */
+    private findOldestFile(): string | undefined {
+        let oldestTime = Number.MAX_SAFE_INTEGER
+        let oldestFile: string | undefined
+
+        for (const [filePath, snapshots] of this.snapshots.entries()) {
+            if (snapshots.length === 0) {
+                continue
+            }
+
+            const oldestSnapshot = snapshots[0]
+            if (oldestSnapshot.timestamp < oldestTime) {
+                oldestTime = oldestSnapshot.timestamp
+                oldestFile = filePath
+            }
+        }
+
+        return oldestFile
+    }
+
+    /**
+     * Gets all snapshots for a specific file
+     * @param filePath The path to the file
+     * @returns Array of snapshots for the file
+     */
+    public getFileSnapshots(filePath: string): FileSnapshot[] {
+        return this.snapshots.get(filePath) || []
+    }
+
+    /**
+     * Gets all tracked files
+     * @returns Array of file paths
+     */
+    public getTrackedFiles(): string[] {
+        return Array.from(this.snapshots.keys())
+    }
+
+    public getTotalSnapshotCount(): number {
+        return Array.from(this.snapshots.values()).reduce((count, snapshots) => count + snapshots.length, 0)
+    }
+
+    public async getSnapshotContent(snapshot: FileSnapshot): Promise<string> {
+        return snapshot.content
+    }
+
+    /**
+     * Generates unified diffs between adjacent snapshots of a file
+     * and between the newest snapshot and the current file content
+     *
+     * @returns Array of SupplementalContext objects containing diffs between snapshots and current content
+     */
+    public async generatePredictionSupplementalContext(): Promise<codewhispererClient.SupplementalContext[]> {
```

Comment:
As called out above

Can you please add a global try-catch + error log for all public methods which will be called for completions code path.